src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -114,67 +116,73 @@ static bool MultiRouterPlannableQuery(Query *query,
 static RelationRestrictionContext * CopyRelationRestrictionContext(
 	RelationRestrictionContext *oldContext);
 static Node * InstantiatePartitionQual(Node *node, void *context);
-static void ErrorIfInsertSelectQueryNotSupported(Query *queryTree,
-												 RangeTblEntry *insertRte,
-												 RangeTblEntry *subqueryRte,
-												 bool allReferenceTables);
-static void ErrorIfMultiTaskRouterSelectQueryUnsupported(Query *query);
-static void ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query,
-														   RangeTblEntry *insertRte,
-														   RangeTblEntry *subqueryRte,
-														   Oid *
-														   selectPartitionColumnTableId);
+static DeferredErrorMessage * InsertSelectQuerySupported(Query *queryTree,
+														 RangeTblEntry *insertRte,
+														 RangeTblEntry *subqueryRte,
+														 bool allReferenceTables);
+static DeferredErrorMessage * MultiTaskRouterSelectQuerySupported(Query *query);
+static DeferredErrorMessage * InsertPartitionColumnMatchesSelect(Query *query,
+																 RangeTblEntry *insertRte,
+																 RangeTblEntry *
+																 subqueryRte,
+																 Oid *
+																 selectPartitionColumnTableId);
 static void AddUninstantiatedEqualityQual(Query *query, Var *targetPartitionColumnVar);
-static void ErrorIfQueryHasModifyingCTE(Query *queryTree);
+static DeferredErrorMessage * ErrorIfQueryHasModifyingCTE(Query *queryTree);
+
 
 /*
- * MultiRouterPlanCreate creates a multi plan for the queries
- * that includes the following:
- *   (i)  modification queries that hit a single shard
- *   (ii) select queries  hat can be executed on a single worker
- *   node and does not require any operations on the master node.
- *   (iii) INSERT INTO .... SELECT queries
- *
- * The function returns NULL if it cannot create the plan for SELECT
- * queries and errors out if it cannot plan the modify queries.
+ * CreateRouterPlan attempts to create a router executor plan for the given
+ * SELECT statement.  If planning fails either NULL is returned, or
+ * ->planningError is set to a description of the failure.
  */
 MultiPlan *
-MultiRouterPlanCreate(Query *originalQuery, Query *query,
-					  RelationRestrictionContext *restrictionContext)
+CreateRouterPlan(Query *originalQuery, Query *query,
+				 RelationRestrictionContext *restrictionContext)
 {
-	MultiPlan *multiPlan = NULL;
+	Assert(EnableRouterExecution);
 
-	bool routerPlannable = MultiRouterPlannableQuery(query, restrictionContext);
-	if (!routerPlannable)
+	if (MultiRouterPlannableQuery(query, restrictionContext))
 	{
-		return NULL;
+		return CreateSingleTaskRouterPlan(originalQuery, query,
+										  restrictionContext);
 	}
 
+	/*
+	 * TODO: Instead have MultiRouterPlannableQuery set an error describing
+	 * why router cannot support the query.
+	 */
+	return NULL;
```

Comment:
Hm. Looked into it, and there's still several sources of NULLs - RouterSelectTask returning NULL, MultiRouterPlannableQuery returning false, ...

I think I'd prefer opening a ticket to clean this stuff up properly, rather than papering over it with a generic message?